File: apps/macos/DNDSync/SyncPorts.swift
import Foundation

protocol SyncLan: AnyObject {
    var onUiState: ((LanUiSnapshot) -> Void)? { get set }
    var onInbound: ((Dndsync_V1_DndState) -> Void)? { get set }
    var onInboundUnpair: ((Dndsync_V1_PairControl) -> Void)? { get set }
    func setPairId(_ pairId: String)
    func start()
    func stop()
    func send(_ state: Dndsync_V1_DndState)
    func sendUnpairNow(_ control: Dndsync_V1_PairControl)
}

extension LanSyncService: SyncLan {}

protocol SyncCloud: AnyObject {
    func postEnvelope(
        baseURL: String,
        pairId: String,
        secret: String,
        envelope: Data
    ) async throws

    func deletePair(baseURL: String, pairId: String, secret: String) async throws
}

extension ForwarderClient: SyncCloud {}

protocol SyncPairing: AnyObject {
    var joined: Bool { get }
    var pairId: String { get }
    var pairSecret: String { get }
    var forwarderURL: String { get }
    func seal(_ plaintext: Data) throws -> Data
    func open(_ ciphertext: Data) -> Data?
    func persistLastSync(unixMs: Int64, on: Bool, sender: String, viaLan: Bool)
    func handleInboundUnpair(_ control: Dndsync_V1_PairControl)
    func noteCloudUnauthorized()
    func noteCloudError(_ message: String)
    func noteCloudSuccess()
}

extension PairSession: SyncPairing {}

File: apps/android/app/src/main/java/com/dndsync/android/sync/SyncPorts.kt
package com.dndsync.android.sync

import com.dndsync.proto.v1.DndState
import com.dndsync.proto.v1.PairControl
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import com.dndsync.android.lan.LanUiState

interface SyncLan {
    val ui: StateFlow<LanUiState>
    val inbound: SharedFlow<DndState>
    val inboundUnpair: SharedFlow<PairControl>
    fun setPairId(pairId: String)
    fun start()
    fun stop()
    /** True when the Mac ACKs this state's `unix_ms` within the attempt budget. */
    suspend fun deliverState(state: DndState): Boolean
    /** Best-effort dial. The caller still posts the cloud unpair. */
    suspend fun deliverUnpair(control: PairControl)
}

interface SyncCloud {
    fun postEnvelope(baseUrl: String, pairId: String, secret: String, envelope: ByteArray)
    fun deletePair(baseUrl: String, pairId: String, secret: String)
}

interface SyncPairing {
    val joined: Boolean
    val pairId: String
    val pairSecret: String
    val forwarderURL: String
    fun seal(plaintext: ByteArray): ByteArray
    fun open(ciphertext: ByteArray): ByteArray?
    fun persistLastSync(unixMs: Long, on: Boolean, sender: String, viaLan: Boolean)
    fun handleInboundUnpair(control: PairControl)
    fun noteCloudUnauthorized()
    fun noteCloudError(message: String)
    fun noteCloudSuccess()
}
